components/raftstore-v2/src/raft/storage.rs

```diff
@@ -240,6 +258,159 @@ impl<ER: RaftEngine> Storage<ER> {
+                "region_id" => self.region_state.get_region().get_id(),
+                "peer_id" => self.peer.get_id(),
+                "err" => ?e,
+            );
+            STORE_SNAPSHOT_VALIDATION_FAILURE_COUNTER.decode.inc();
+            return false;
+        }
+        let snap_epoch = snap_data.get_region().get_region_epoch();
+        let latest_epoch = self.region().get_region_epoch();
+        if snap_epoch.get_conf_ver() < latest_epoch.get_conf_ver() {
+            info!(
+                self.logger(),
+                "snapshot epoch is stale";
+                "region_id" => self.region_state.get_region().get_id(),
+                "peer_id" => self.peer.get_id(),
+                "snap_epoch" => ?snap_epoch,
+                "latest_epoch" => ?latest_epoch,
+            );
+            STORE_SNAPSHOT_VALIDATION_FAILURE_COUNTER.epoch.inc();
+            return false;
+        }
+
+        true
+    }
+
+    /// Gets a snapshot. Returns `SnapshotTemporarilyUnavailable` if there is no
+    /// unavailable snapshot.
+    pub fn snapshot(&self, request_index: u64, to: u64) -> raft::Result<Snapshot> {
+        let mut snap_state = self.snap_state.borrow_mut();
+
+        let mut tried = false;
+        let mut last_canceled = false;
+        if let SnapState::Generating {
+            ref canceled,
+            ref receiver,
+            ..
+        } = *snap_state
+        {
+            tried = true;
+            last_canceled = canceled.load(Ordering::SeqCst);
+            match receiver.try_recv() {
+                Err(TryRecvError::Empty) => {
+                    return Err(raft::Error::Store(
+                        raft::StorageError::SnapshotTemporarilyUnavailable,
+                    ));
+                }
+                Ok(s) if !last_canceled => {
+                    *snap_state = SnapState::Relax;
+                    if self.validate_snap(&s, request_index) {
+                        return Ok(s);
+                    }
+                }
+                Err(TryRecvError::Disconnected) | Ok(_) => {
+                    *snap_state = SnapState::Relax;
+                    warn!(
+                        self.logger(),
+                        "failed to try generating snapshot";
+                        "region_id" => self.region().get_id(),
+                        "peer_id" => self.peer().get_id(),
+                        "request_peer" => to,
+                    );
+                }
+            }
+        }
+
+        if SnapState::Relax != *snap_state {
+            panic!(
+                "[region {}] unexpected state: {:?}",
+                self.region().get_id(),
+                *snap_state
+            );
+        }
+
+        if !tried || !last_canceled {
+            STORE_SNAPSHOT_RETIRES_COUNTER.inc();
+        }
+
+        info!(
+            self.logger(),
+            "requesting snapshot";
+            "region_id" => self.region().get_id(),
```

Comment:
All IDs are included in the logger.

Reply:
Do you mean the logger will init with the head like `[region xxx][1,2,3]`?  I don't see that.

Reply:
> Do you mean the logger will init with the head like [region xxx][1,2,3]?

Yes, check

https://github.com/tikv/tikv/blob/94c8aa5c3c5335ea99281bea7c6c79f61cb201de/components/raftstore-v2/src/raft/storage.rs#L202